File: solutions/fortran/collatz-conjecture/1/collatz_conjecture.f90

module collatz_conjecture
  implicit none
contains

  integer function steps(i)
    integer :: i, j

    steps = -1
    j = i

    if (j >= 1) then
      steps = 0
      do while (j > 1)
        if (mod(j, 2) == 0) then
          steps = steps + 1;
          j = j / 2;
        else
          steps = steps + 1;
          j = (3 * j) + 1;
        end if
      end do
    end if
  end function

end module

File: solutions/cpp/collatz-conjecture/1/collatz_conjecture.cpp
#include "collatz_conjecture.h"

#include <stdexcept>

namespace collatz_conjecture
{

    int steps(int number)
    {
        if (number <= 0)
            throw std::domain_error{"Invalid starting number"};

        int steps = 0;
        while (number != 1)
        {
            if (number % 2 == 0)
            {
                number /= 2;
            }
            else
            {
                number = 3 * number + 1;
            }
            steps++;
        }

        return steps;
    }

} // namespace collatz_conjecture
